### PR TITLE
Migrate S4165 (Redundant Assignments) to eslint implementation

### LINKED ISSUE #2332: Migrate S4165 (Redundant Assignments) to eslint implementation
https://jira.sonarsource.com/browse/RSPEC-4165

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #2332

### CODE REVIEW

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
There seems to be a potentially problematic FP which I'm not sure to understand the exact root cause of. 
As I'm still trying to understand the overall implementation, I let you have a look at it as well as the other minor comments I had so far.

File: its/ruling/src/test/expected/js/p5.js/javascript-S4165.json [hunk 1]
```diff
@@ -1,34 +1,30 @@
...
 ],
 'p5.js:src/core/error_helpers.js':[
 419,
 ],
+'p5.js:src/image/image.js':[
+233,
+],
+'p5.js:src/typography/p5.Font.js':[
+718,
+719,
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply its/ruling/src/test/expected/js/p5.js/javascript-S4165.json:13]: Indeed, I forgot to update `assignedValuesMap` inside `checkSegment` function. 
Thanks!

File: its/ruling/src/test/expected/js/jshint/javascript-S4165.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+'jshint:tests/unit/fixtures/redef-es6.js':[
+270,
+],
+'jshint:tests/unit/fixtures/scope.js':[
+6,
+17,
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply its/ruling/src/test/expected/js/jshint/javascript-S4165.json:7]: I decided to exclude those kind of issues, thanks!

File: eslint-bridge/src/rules/no-redundant-assignments.ts [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+import CodePathSegment = Rule.CodePathSegment;
+import {
+  areEquivalent,
+  reachingDefinitions,
+  ReachingDefinitions,
+  resolveAssignedValues,
+} from './reachingDefinitions';
+
+const message = (name: string) =>
+  `Review this useless assignment: "${name}" already holds the assigned value along all execution paths.`;
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply eslint-bridge/src/rules/no-redundant-assignments.ts:38]: good point! I think we use `redundant`

File: javascript-checks/src/main/java/org/sonar/javascript/checks/RedundantAssignmentCheck.java [hunk 4]
```diff
@@ -32,42 +33,10 @@
 
 @JavaScriptRule
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply javascript-checks/src/main/java/org/sonar/javascript/checks/RedundantAssignmentCheck.java:34]: Indeed, I forgot to activate it for TS!

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
I think there is a small bug in `getVariableFromIdentifier` that somehow doesn't impact tests. Other than that, LGTM.

File: eslint-bridge/src/rules/reachingDefinitions.ts [hunk 5]
```diff
@@ -175,3 +169,13 @@ function equals(ps1: Map<Variable, VariableValue>, ps2: Map<Variable, VariableVa
 function setEquals<T>(a: Set<T>, b: Set<T>): boolean {
   return a.size === b.size && [...a].every(e => b.has(e));
 }
+
+export function getVariableFromIdentifier(identifier: estree.Identifier, scope: Scope.Scope) {
+  let variable;
+  let sc: Scope.Scope | null = scope;
+  while (variable == null && sc != null) {
+    variable = scope.variables.find(value => value.name === identifier.name);
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply eslint-bridge/src/rules/reachingDefinitions.ts:177]: indeed, it wouldn't resolve variables defined in an outer scope. I will add a UT for it! Thanks!

**Pull request merged**